File: 04-Design-Principles/src/go/movement.go
package main

import "fmt"

type MovementBehavior interface {
	Move(name string)
}

type WalkMovement struct{}
func (WalkMovement) Move(name string) { fmt.Printf("%s walks.\n", name) }

type FlyMovement struct{}
func (FlyMovement) Move(name string) { fmt.Printf("%s flies!\n", name) }

type TeleportMovement struct{}
func (TeleportMovement) Move(name string) { fmt.Printf("%s teleports!\n", name) }

File: 04-Design-Principles/src/rust/movement.rs
pub trait MovementBehavior {
    fn do_move(&self, name: &str);
}

pub struct WalkMovement;
impl MovementBehavior for WalkMovement {
    fn do_move(&self, name: &str) { println!("{} walks.", name); }
}

pub struct FlyMovement;
impl MovementBehavior for FlyMovement {
    fn do_move(&self, name: &str) { println!("{} flies!", name); }
}

pub struct TeleportMovement;
impl MovementBehavior for TeleportMovement {
    fn do_move(&self, name: &str) { println!("{} teleports!", name); }
}
